src/tools/storage-perf/StoragePerfTool.cpp

```diff
@@ -196,91 +179,121 @@ class Perf {
     }
 
     std::vector<storage::cpp2::Edge> genEdges() {
-        static int32_t size = sizeof(slots_) / sizeof(int32_t);
         std::vector<storage::cpp2::Edge> edges;
         static VertexID vId = FLAGS_min_vertex_id;
-        position = (position + 1) % size;
-        for (int32_t i = 0; i< slots_[position]; i++) {
-            storage::cpp2::Edge edge;
-            storage::cpp2::EdgeKey eKey;
-            eKey.set_src(vId);
-            eKey.set_edge_type(edgeType_);
-            eKey.set_dst(vId + 1);
-            eKey.set_ranking(0);
-            edge.set_key(std::move(eKey));
-            auto props = genData(FLAGS_size);
-            edge.set_props(std::move(props));
-            edges.emplace_back(std::move(edge));
-        }
+        storage::cpp2::Edge edge;
+        storage::cpp2::EdgeKey eKey;
+        eKey.set_src(vId);
+        eKey.set_edge_type(edgeType_);
+        eKey.set_dst(vId + 1);
+        eKey.set_ranking(0);
+        edge.set_key(std::move(eKey));
+        auto props = genData(FLAGS_size);
+        edge.set_props(std::move(props));
+        edges.emplace_back(std::move(edge));
         return edges;
     }
 
     void getNeighborsTask() {
         auto* evb = threadPool_->getEventBase();
-        std::vector<EdgeType> e(edgeType_);
-        auto f =
-          client_->getNeighbors(spaceId_, randomVertices(), std::move(e), "", randomCols())
+        auto tokens = tokenBucket_.consumeOrDrain(FLAGS_concurrency, FLAGS_qps, FLAGS_concurrency);
+        for (auto i = 0; i < tokens; i++) {
```

Comment:
I have a question,  when does the value of this `tokens ` not equal `FLAGS_concurrency`?

Reply:
It depends the token rate.

Reply:
The third parameter should be the qps?
> A token bucket stores up to a fixed number of tokens (the burst size)

Reply:
Never mind